File: tests/md_catTest.php
<?php

use FixRefs\Tests\MyFunctionTest;

use function WpRefs\MdCat\add_Translated_from_MDWiki;

class md_catTest extends MyFunctionTest
{
    public function testEquals()
    {
        $text = "[[Kategorija:Translated from MDWiki]]";

        $result = add_Translated_from_MDWiki($text, "hr");

        $this->assertEquals($text, $result);
    }

    public function testSkipLangsIt()
    {
        $text = "This is a sample text";
        $result = add_Translated_from_MDWiki($text, "it");
        $this->assertEquals($text, $result);
    }

    // Test case: Appends category when conditions are met
    public function testAppendsCategoryWhenConditionsMet()
    {
        $text = "This is a sample text";
        $result = add_Translated_from_MDWiki($text, "fr");

        $expected = "This is a sample text\n[[Catégorie:Traduit de MDWiki]]\n";
        $this->assertEquals($expected, $result);
    }

    // Test case: Doesn't append when category is empty
    public function testDoesNotAppendWhenCategoryEmpty()
    {
        $text = "This is a sample text";
        $result = add_Translated_from_MDWiki($text, "fr");
        $expected = "This is a sample text\n[[Catégorie:Traduit de MDWiki]]\n";
        $this->assertEquals($expected, $result);
    }

    // Test case: Doesn't append when category already exists
    public function testDoesNotAppendWhenCategoryExists()
    {
        $category = "[[Category:Translated from MDWiki (de)]]";
        $text = "This is a sample text\n" . $category;

        $result = add_Translated_from_MDWiki($text, "de");

        $this->assertEquals($text, $result);
    }

    // Test case: Doesn't append when fallback category exists
    public function testDoesNotAppendWhenFallbackCategoryExists()
    {
        $text = "This is a sample text\n[[Category:Translated from MDWiki]]";

        $result = add_Translated_from_MDWiki($text, "es");

        $this->assertEquals($text, $result);
    }

    // Test case: Appends when similar but different category exists
    public function testAppendsWhenSimilarCategoryExists()
    {
        $text = "This is a sample text\n[[Category:Translated from MDWiki]]\n";

        $result = add_Translated_from_MDWiki($text, "ja");

        $this->assertEquals($text, $result);
    }

    // Test case: Handles multiple newlines correctly
    public function testHandlesMultipleNewlines()
    {
        $text = "This is a sample text\n\n";

        $result = add_Translated_from_MDWiki($text, "ru");

        $expected = "This is a sample text\n\n\n[[Категория:Статьи, переведённые с MDWiki]]\n";
        $this->assertEquals($expected, $result);
    }

    public function testLangs()
    {
        $langs = [
            "ur" => "زمرہ:ایم ڈی وکی سے ترجمہ شدہ",
        ];
        foreach ($langs as $lang => $cat) {
            $text_no_cat = "This is a sample text\n\n";
            $expected = "{$text_no_cat}\n[[{$cat}]]\n";
            $result = add_Translated_from_MDWiki($text_no_cat, $lang);
            $this->assertEqualCompare($expected, $text_no_cat, $result);
            // ---
            $text_with_cat = "This is a sample text\n\n[[{$cat}]]\n";
            $result = add_Translated_from_MDWiki($text_with_cat, $lang);
            $this->assertEquals($text_with_cat, $result);
            // ---
            $text_with_cat2 = "This is a sample text\n\n[[category:Translated_from_MDWiki]]\n";
            $result = add_Translated_from_MDWiki($text_with_cat2, $lang);
            $this->assertEquals($text_with_cat2, $result);
            // ---
        }
    }
}

File: tests/md_catTest.py
from __future__ import annotations

from typing import Callable, Dict

import pytest

from src import md_cat
from tests.conftest import assert_equal_compare


@pytest.fixture
def mdwiki_categories(monkeypatch: pytest.MonkeyPatch) -> Callable[[Dict[str, Dict[str, str]]], None]:
    def apply(mapping: Dict[str, Dict[str, str]]) -> None:
        original = md_cat.get_cats
        if hasattr(original, "cache_clear"):
            original.cache_clear()

        def _get_cats() -> Dict[str, Dict[str, str]]:
            return mapping

        monkeypatch.setattr(md_cat, "get_cats", _get_cats)

    return apply


class md_catTest:
    def testEquals(self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]) -> None:
        mdwiki_categories({"hrwiki": {"title": "Category:Translated from MDWiki"}})
        text = "[[Kategorija:Translated from MDWiki]]"
        assert md_cat.add_Translated_from_MDWiki(text, "hr") == text

    def testSkipLangsIt(self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]) -> None:
        mdwiki_categories({})
        text = "This is a sample text"
        assert md_cat.add_Translated_from_MDWiki(text, "it") == text

    def testAppendsCategoryWhenConditionsMet(
        self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]
    ) -> None:
        mdwiki_categories({"frwiki": {"title": "Catégorie:Traduit de MDWiki"}})
        text = "This is a sample text"
        expected = "This is a sample text\n[[Catégorie:Traduit de MDWiki]]\n"
        assert md_cat.add_Translated_from_MDWiki(text, "fr") == expected

    def testDoesNotAppendWhenCategoryEmpty(
        self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]
    ) -> None:
        mdwiki_categories({"frwiki": {"title": "Catégorie:Traduit de MDWiki"}})
        text = "This is a sample text"
        expected = "This is a sample text\n[[Catégorie:Traduit de MDWiki]]\n"
        assert md_cat.add_Translated_from_MDWiki(text, "fr") == expected

    def testDoesNotAppendWhenCategoryExists(
        self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]
    ) -> None:
        mdwiki_categories({"dewiki": {"title": "Category:Translated from MDWiki (de)"}})
        category = "[[Category:Translated from MDWiki (de)]]"
        text = f"This is a sample text\n{category}"
        assert md_cat.add_Translated_from_MDWiki(text, "de") == text

    def testDoesNotAppendWhenFallbackCategoryExists(
        self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]
    ) -> None:
        mdwiki_categories({"eswiki": {"title": "Category:Translated from MDWiki"}})
        text = "This is a sample text\n[[Category:Translated from MDWiki]]"
        assert md_cat.add_Translated_from_MDWiki(text, "es") == text

    def testAppendsWhenSimilarCategoryExists(
        self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]
    ) -> None:
        mdwiki_categories({"jawiki": {"title": "Category:Translated from MDWiki"}})
        text = "This is a sample text\n[[Category:Translated from MDWiki]]\n"
        assert md_cat.add_Translated_from_MDWiki(text, "ja") == text

    def testHandlesMultipleNewlines(
        self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]
    ) -> None:
        mdwiki_categories({"ruwiki": {"title": "Категория:Статьи, переведённые с MDWiki"}})
        text = "This is a sample text\n\n"
        expected = "This is a sample text\n\n\n[[Категория:Статьи, переведённые с MDWiki]]\n"
        assert md_cat.add_Translated_from_MDWiki(text, "ru") == expected

    def testLangs(self, mdwiki_categories: Callable[[Dict[str, Dict[str, str]]], None]) -> None:
        langs = {"urwiki": {"title": "زمرہ:ایم ڈی وکی سے ترجمہ شدہ"}}
        mdwiki_categories(langs)
        text_no_cat = "This is a sample text\n\n"
        expected = "This is a sample text\n\n\n[[زمرہ:ایم ڈی وکی سے ترجمہ شدہ]]\n"
        result = md_cat.add_Translated_from_MDWiki(text_no_cat, "ur")
        assert_equal_compare(expected, text_no_cat, result)

        text_with_cat = "This is a sample text\n\n[[زمرہ:ایم ڈی وکی سے ترجمہ شدہ]]\n"
        assert md_cat.add_Translated_from_MDWiki(text_with_cat, "ur") == text_with_cat

        text_with_fallback = "This is a sample text\n\n[[category:Translated_from_MDWiki]]\n"
        assert md_cat.add_Translated_from_MDWiki(text_with_fallback, "ur") == text_with_fallback
